File: src/python/pulse/howto/HowTo_MechanicalVentilator.py
# Distributed under the Apache License, Version 2.0.
# See accompanying NOTICE file for details.

from pulse.engine.PulseEngine import PulseEngine
from pulse.cdm.mechanical_ventilator_actions import SEMechanicalVentilatorConfiguration, \
                                                    SEMechanicalVentilatorContinuousPositiveAirwayPressure, \
                                                    SEMechanicalVentilatorPressureControl, \
                                                    SEMechanicalVentilatorVolumeControl, \
                                                    SEMechanicalVentilatorHold, \
                                                    SEMechanicalVentilatorLeak, \
                                                    eMechanicalVentilator_PressureControlMode, \
                                                    eMechanicalVentilator_VolumeControlMode
from pulse.cdm.scalars import FrequencyUnit, PressureUnit, PressureTimePerVolumeUnit, \
                              TimeUnit, VolumeUnit, VolumePerPressureUnit, VolumePerTimeUnit
from pulse.cdm.mechanical_ventilator import eSwitch, eDriverWaveform
from pulse.cdm.engine import SEDataRequest, SEDataRequestManager
from pulse.cdm.patient_actions import SEDyspnea


def HowTo_MechanicalVentilator():
    pulse = PulseEngine()
    pulse.set_log_filename("./test_results/howto/HowTo_MechanicalVentilator.py.log")
    pulse.log_to_console(True)

    data_requests = [
        SEDataRequest.create_physiology_request("RespirationRate", unit=FrequencyUnit.Per_min),
        # Ventilator Monitor Data
        SEDataRequest.create_mechanical_ventilator_request("RespirationRate", unit=FrequencyUnit.Per_min)]

    data_mgr = SEDataRequestManager(data_requests)
    data_mgr.set_results_filename("./test_results/howto/HowTo_MechanicalVentilator.py.csv")

    # NOTE: No data requests are being provided, so Pulse will return the default vitals data
    if not pulse.serialize_from_file("./states/StandardMale@0s.json", data_mgr):
        print("Unable to load initial state file")
        return

    # Get default data at time 0s from the engine
    results = pulse.pull_data()
    data_mgr.to_console(results)

    # Give the patient Dyspnea
    dsypnea = SEDyspnea()
    dsypnea.set_comment("Patient's dsypnea occurs")
    dsypnea.get_severity().set_value(1.0)
    pulse.process_action(dsypnea)
    pulse.advance_time_s(10)
    # Get the values of the data you requested at this time
    results = pulse.pull_data()
    # And write it out to the console
    data_mgr.to_console(results)

    # We have action support for several commonly used ventilator modes
    # Pulse is not limited to these modes, These modes are designed for simple understanding
    # Our implementation supports any ventilator mode, you will just need to translate the user facing inputs
    # to a timing/control profile using our configuration settings
    # These modes are internally converted into a configuration setting,
    # You can retrieve the resulting settings action and use it as a basis for more configurations if you want.
    # For example, if you wanted to lengthen the InspirationPatientTriggerFlow of the mode

    cpap = SEMechanicalVentilatorContinuousPositiveAirwayPressure()
    cpap.set_connection(eSwitch.On)
    cpap.get_fraction_inspired_oxygen().set_value(0.21)
    cpap.get_delta_pressure_support().set_value(8.0, PressureUnit.cmH2O)
    cpap.get_positive_end_expired_pressure().set_value(5.0, PressureUnit.cmH2O)
    cpap.get_slope().set_value(0.2, TimeUnit.s)
    pulse.process_action(cpap)
    pulse.advance_time_s(10)
    # Get the values of the data you requested at this time
    results = pulse.pull_data()
    # And write it out to the console
    data_mgr.to_console(results)

    pc_ac = SEMechanicalVentilatorPressureControl()
    pc_ac.set_connection(eSwitch.On)
    pc_ac.set_mode(eMechanicalVentilator_PressureControlMode.AssistedControl)
    pc_ac.get_fraction_inspired_oxygen().set_value(0.21)
    pc_ac.get_inspiratory_period().set_value(1.0, TimeUnit.s)
    pc_ac.get_inspiratory_pressure().set_value(13.0, PressureUnit.cmH2O)
    pc_ac.get_positive_end_expired_pressure().set_value(5.0, PressureUnit.cmH2O)
    pc_ac.get_respiration_rate().set_value(12.0, FrequencyUnit.Per_min)
    pc_ac.get_slope().set_value(0.1, TimeUnit.s)
    pulse.process_action(pc_ac)
    pulse.advance_time_s(10)
    # Get the values of the data you requested at this time
    results = pulse.pull_data()
    # And write it out to the console
    data_mgr.to_console(results)

    vc_ac = SEMechanicalVentilatorVolumeControl()
    vc_ac.set_connection(eSwitch.On)
    vc_ac.set_mode(eMechanicalVentilator_VolumeControlMode.AssistedControl)
    vc_ac.get_flow().set_value(50.0, VolumePerTimeUnit.L_Per_min)
    vc_ac.get_fraction_inspired_oxygen().set_value(0.21)
    vc_ac.get_inspiratory_period().set_value(1.0, TimeUnit.s)
    vc_ac.get_positive_end_expired_pressure().set_value(5.0, PressureUnit.cmH2O)
    vc_ac.get_respiration_rate().set_value(12.0, FrequencyUnit.Per_min)
    vc_ac.get_tidal_volume().set_value(600.0, VolumeUnit.mL)
    pulse.process_action(vc_ac)
    pulse.advance_time_s(10)
    # Get the values of the data you requested at this time
    results = pulse.pull_data()
    # And write it out to the console
    data_mgr.to_console(results)

    # Here is an example of programming a custom ventilator mode
    mechanical_ventilator_cfg = SEMechanicalVentilatorConfiguration()
    mechanical_ventilator_cfg.set_comment("Attach the mechanical ventilator")
    # You could just point to an mechanical ventilator configuration file if you want (no examples)
    # mechanical_ventilator.set_settings_file("/path/to/file")
    mechanical_ventilator = mechanical_ventilator_cfg.get_settings()
    mechanical_ventilator.set_connection(eSwitch.On)
    mechanical_ventilator.set_inspiration_waveform(eDriverWaveform.Square)
    mechanical_ventilator.set_expiration_waveform(eDriverWaveform.Square)
    mechanical_ventilator.get_peak_inspiratory_pressure().set_value(21.0, PressureUnit.cmH2O)
    mechanical_ventilator.get_positive_end_expired_pressure().set_value(10.0, PressureUnit.cmH2O)
    mechanical_ventilator.get_fraction_inspired_gas("Oxygen").get_fraction_amount().set_value(0.5)
    respiration_rate_per_min = 20.0
    ie_ratio = 0.5
    # Translate ventilator settings
    total_period_s = 60.0 / respiration_rate_per_min
    inspiratory_period_s = ie_ratio * total_period_s / (1 + ie_ratio)
    expiratory_period_s = total_period_s - inspiratory_period_s
    mechanical_ventilator.get_inspiration_machine_trigger_time().set_value(expiratory_period_s, TimeUnit.s)
    mechanical_ventilator.get_expiration_cycle_time().set_value(inspiratory_period_s, TimeUnit.s)
    pulse.process_action(mechanical_ventilator_cfg)
    # Advance some time and print out the vitals
    pulse.advance_time_s(10)
    # Get the values of the data you requested at this time
    results = pulse.pull_data()
    # And write it out to the console
    data_mgr.to_console(results)

    # You can also perform a hold
    hold = SEMechanicalVentilatorHold()
    hold.set_state(eSwitch.On)
    pulse.process_action(hold)
    # Advance some time and print out the vitals
    pulse.advance_time_s(5)
    # Get the values of the data you requested at this time
    results = pulse.pull_data()
    # And write it out to the console
    data_mgr.to_console(results)
    hold.set_state(eSwitch.Off)
    pulse.process_action(hold)
    # Advance some time and print out the vitals
    pulse.advance_time_s(5)
    # Get the values of the data you requested at this time
    results = pulse.pull_data()
    # And write it out to the console
    data_mgr.to_console(results)

    # A leak can be specified
    leak = SEMechanicalVentilatorLeak()
    leak.get_severity().set_value(0.5)
    pulse.process_action(leak)
    # Advance some time and print out the vitals
    pulse.advance_time_s(5)
    # Get the values of the data you requested at this time
    results = pulse.pull_data()
    # And write it out to the console
    data_mgr.to_console(results)
    leak.get_severity().set_value(0.0)  # Turn off the leak
    pulse.process_action(leak)
    # Advance some time and print out the vitals
    pulse.advance_time_s(5)
    # Get the values of the data you requested at this time
    results = pulse.pull_data()
    # And write it out to the console
    data_mgr.to_console(results)


HowTo_MechanicalVentilator()

File: src/java/com/kitware/pulse/howto/HowTo_MechanicalVentilator.java
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/
package com.kitware.pulse.howto;

import java.util.List;

import com.kitware.pulse.cdm.bind.Enums.eDriverWaveform;
import com.kitware.pulse.cdm.bind.Enums.eSwitch;
import com.kitware.pulse.cdm.bind.MechanicalVentilatorActions.*;
import com.kitware.pulse.cdm.datarequests.SEDataRequestManager;
import com.kitware.pulse.cdm.patient.actions.SEDyspnea;
import com.kitware.pulse.cdm.properties.CommonUnits.*;
import com.kitware.pulse.cdm.properties.SEScalarTime;
import com.kitware.pulse.cdm.substance.SESubstanceFraction;
import com.kitware.pulse.cdm.system.equipment.mechanical_ventilator.SEMechanicalVentilatorSettings;
import com.kitware.pulse.cdm.system.equipment.mechanical_ventilator.actions.*;
import com.kitware.pulse.engine.PulseEngine;
import com.kitware.pulse.utilities.JNIBridge;
import com.kitware.pulse.utilities.Log;
import com.kitware.pulse.utilities.LogListener;

public class HowTo_MechanicalVentilator
{
  protected static class MyListener extends LogListener
  {
    protected MyListener()
    {
      super();
      listen(false);
    }    
    @Override public void handleDebug(String msg) { Log.debug(msg); }
    @Override public void handleInfo(String msg)  { Log.info(msg); }
    @Override public void handleWarn(String msg)  { Log.warn(msg); }
    @Override public void handleError(String msg) { Log.error(msg); }
    @Override public void handleFatal(String msg) { Log.fatal(msg); }
  }
  
  public static void main(String[] args)
  {
    List<Double> data_values;
    SEScalarTime time = new SEScalarTime(1,TimeUnit.min);
    
    JNIBridge.initialize();
    PulseEngine pulse = new PulseEngine();
    pulse.setLogListener(new MyListener());
    pulse.setLogFilename("./test_results/howto/HowTo_MechanicalVentilator.java.log");
    
    SEDataRequestManager dataRequests = new SEDataRequestManager();// Vitals Monitor Data
    dataRequests.createPhysiologyDataRequest("HeartRate", FrequencyUnit.Per_min);
    dataRequests.createPhysiologyDataRequest("ArterialPressure", PressureUnit.mmHg);
    dataRequests.createPhysiologyDataRequest("MeanArterialPressure", PressureUnit.mmHg);
    dataRequests.createPhysiologyDataRequest("SystolicArterialPressure", PressureUnit.mmHg);
    dataRequests.createPhysiologyDataRequest("DiastolicArterialPressure", PressureUnit.mmHg);
    dataRequests.createPhysiologyDataRequest("OxygenSaturation");
    dataRequests.createPhysiologyDataRequest("EndTidalCarbonDioxidePressure", PressureUnit.mmHg);
    dataRequests.createPhysiologyDataRequest("RespirationRate", FrequencyUnit.Per_min);
    dataRequests.createPhysiologyDataRequest("SkinTemperature", TemperatureUnit.C);
    dataRequests.createGasCompartmentDataRequest("Carina", "CarbonDioxide", "PartialPressure", PressureUnit.mmHg);
    dataRequests.createPhysiologyDataRequest("BloodVolume", VolumeUnit.mL);
    dataRequests.createECGDataRequest("Lead3ElectricPotential", ElectricPotentialUnit.mV);
    // Ventilator Monitor Data
    dataRequests.createMechanicalVentilatorDataRequest("AirwayPressure", PressureUnit.cmH2O);
    dataRequests.createMechanicalVentilatorDataRequest("EndTidalCarbonDioxideFraction");
    dataRequests.createMechanicalVentilatorDataRequest("EndTidalCarbonDioxidePressure", PressureUnit.cmH2O);
    dataRequests.createMechanicalVentilatorDataRequest("EndTidalOxygenFraction");
    dataRequests.createMechanicalVentilatorDataRequest("EndTidalOxygenPressure", PressureUnit.cmH2O);
    dataRequests.createMechanicalVentilatorDataRequest("ExpiratoryFlow", VolumePerTimeUnit.L_Per_s);
    dataRequests.createMechanicalVentilatorDataRequest("ExpiratoryTidalVolume", VolumeUnit.L);
    dataRequests.createMechanicalVentilatorDataRequest("InspiratoryExpiratoryRatio");
    dataRequests.createMechanicalVentilatorDataRequest("InspiratoryFlow", VolumePerTimeUnit.L_Per_s);
    dataRequests.createMechanicalVentilatorDataRequest("InspiratoryTidalVolume", VolumeUnit.L);
    dataRequests.createMechanicalVentilatorDataRequest("IntrinsicPositiveEndExpiredPressure", PressureUnit.cmH2O);
    dataRequests.createMechanicalVentilatorDataRequest("LeakFraction");
    dataRequests.createMechanicalVentilatorDataRequest("MeanAirwayPressure", PressureUnit.cmH2O);
    dataRequests.createMechanicalVentilatorDataRequest("PeakInspiratoryPressure", PressureUnit.cmH2O);
    dataRequests.createMechanicalVentilatorDataRequest("PlateauPressure", PressureUnit.cmH2O);
    dataRequests.createMechanicalVentilatorDataRequest("PositiveEndExpiratoryPressure", PressureUnit.cmH2O);
    dataRequests.createMechanicalVentilatorDataRequest("RespirationRate", FrequencyUnit.Per_min);
    dataRequests.createMechanicalVentilatorDataRequest("TidalVolume", VolumeUnit.L);
    dataRequests.createMechanicalVentilatorDataRequest("TotalLungVolume", VolumeUnit.L);
    dataRequests.createMechanicalVentilatorDataRequest("TotalPulmonaryVentilation", VolumePerTimeUnit.L_Per_s);
    
    pulse.serializeFromFile("./states/StandardMale@0s.json", dataRequests);
    
    SEDyspnea dyspnea = new SEDyspnea();
    dyspnea.getSeverity().setValue(1.0);
    pulse.processAction(dyspnea);
    
 // We have action support for several commonly used ventilator modes
    // Pulse is not limited to these modes, These modes are designe for simple understanding
    // Our implementation supports any ventilator mode, you will just need to translate the user facing inputs
    // to a timing/control profile using our configuration settings
    // These modes are internally converted into a configuration setting,
    // You can retrieve the resulting settings action and use it as a basis for more configurations if you want.
    // For example, if you wanted to lengthen the InspirationPatientTriggerFlow of the mode

    SEMechanicalVentilatorContinuousPositiveAirwayPressure cpap = new SEMechanicalVentilatorContinuousPositiveAirwayPressure();
    cpap.setConnection(eSwitch.On);
    cpap.getFractionInspiredOxygen().setValue(0.21);
    cpap.getDeltaPressureSupport().setValue(10.0, PressureUnit.cmH2O);
    cpap.getPositiveEndExpiredPressure().setValue(5.0, PressureUnit.cmH2O);
    cpap.getSlope().setValue(0.2, TimeUnit.s);
    pulse.processAction(cpap);
    pulse.advanceTime_s(10);
    // get the values of the data you requested at this time
    data_values = pulse.pullData();
    // And write it out to the console
    dataRequests.writeData(data_values);

    SEMechanicalVentilatorPressureControl pc_ac = new SEMechanicalVentilatorPressureControl();
    pc_ac.setConnection(eSwitch.On);
    pc_ac.setMode(MechanicalVentilatorPressureControlData.eMode.AssistedControl);
    pc_ac.getFractionInspiredOxygen().setValue(0.21);
    pc_ac.getInspiratoryPeriod().setValue(1.0,TimeUnit.s);
    pc_ac.getInspiratoryPressure().setValue(19.0, PressureUnit.cmH2O);
    pc_ac.getPositiveEndExpiredPressure().setValue(5.0, PressureUnit.cmH2O);
    pc_ac.getRespirationRate().setValue(12.0, FrequencyUnit.Per_min);
    pc_ac.getSlope().setValue(0, TimeUnit.s);
    pulse.processAction(pc_ac);
    pulse.advanceTime_s(10);
    // get the values of the data you requested at this time
    data_values = pulse.pullData();
    // And write it out to the console
    dataRequests.writeData(data_values);

    SEMechanicalVentilatorVolumeControl vc_ac = new SEMechanicalVentilatorVolumeControl();
    vc_ac.setConnection(eSwitch.On);
    vc_ac.setMode(MechanicalVentilatorVolumeControlData.eMode.AssistedControl);
    vc_ac.getFlow().setValue(60.0, VolumePerTimeUnit.L_Per_min);
    vc_ac.getFractionInspiredOxygen().setValue(0.21);
    vc_ac.getInspiratoryPeriod().setValue(1.0, TimeUnit.s);
    vc_ac.getPositiveEndExpiredPressure().setValue(5.0, PressureUnit.cmH2O);
    vc_ac.getRespirationRate().setValue(12.0, FrequencyUnit.Per_min);
    vc_ac.getTidalVolume().setValue(900.0, VolumeUnit.mL);
    pulse.processAction(vc_ac);
    pulse.advanceTime_s(10);
    // get the values of the data you requested at this time
    data_values = pulse.pullData();
    // And write it out to the console
    dataRequests.writeData(data_values);

    // Here is an example of programming a custom ventilator mode
    SEMechanicalVentilatorConfiguration mv_config = new SEMechanicalVentilatorConfiguration();
    SEMechanicalVentilatorSettings mv = mv_config.getSettings();
    mv.setConnection(eSwitch.On);
    mv.setInspirationWaveform(eDriverWaveform.Square);
    mv.setExpirationWaveform(eDriverWaveform.Square);
    mv.getPeakInspiratoryPressure().setValue(21.0, PressureUnit.cmH2O);
    mv.getPositiveEndExpiredPressure().setValue(10.0, PressureUnit.cmH2O);
    SESubstanceFraction fractionFiO2 = mv.getFractionInspiredGas("Oxygen");
    fractionFiO2.getAmount().setValue(0.5);
    double respirationRate_per_min = 20.0;
    double IERatio = 0.5;
    // Translate ventilator settings
    double totalPeriod_s = 60.0 / respirationRate_per_min;
    double inspiratoryPeriod_s = IERatio * totalPeriod_s / (1 + IERatio);
    double expiratoryPeriod_s = totalPeriod_s - inspiratoryPeriod_s;
    mv.getInspirationMachineTriggerTime().setValue(expiratoryPeriod_s, TimeUnit.s);
    mv.getExpirationCycleTime().setValue(inspiratoryPeriod_s, TimeUnit.s);
    pulse.processAction(mv_config);
    // Advance some time and print out the vitals
    pulse.advanceTime_s(10);
    // get the values of the data you requested at this time
    data_values = pulse.pullData();
    // And write it out to the console
    dataRequests.writeData(data_values);

    // You can also perform a hold
    SEMechanicalVentilatorHold hold = new SEMechanicalVentilatorHold();
    hold.setState(eSwitch.On);
    pulse.processAction(hold);
    // Advance some time and print out the vitals
    pulse.advanceTime_s(5);
    // get the values of the data you requested at this time
    data_values = pulse.pullData();
    // And write it out to the console
    dataRequests.writeData(data_values);
    hold.setState(eSwitch.Off);
    pulse.processAction(hold);
    // Advance some time and print out the vitals
    pulse.advanceTime_s(5);
    // get the values of the data you requested at this time
    data_values = pulse.pullData();
    // And write it out to the console
    dataRequests.writeData(data_values);

    // A leak can be specified
    SEMechanicalVentilatorLeak leak = new SEMechanicalVentilatorLeak();
    leak.getSeverity().setValue(0.5);
    pulse.processAction(leak);
    // Advance some time and print out the vitals
    pulse.advanceTime_s(5);
    // get the values of the data you requested at this time
    data_values = pulse.pullData();
    // And write it out to the console
    dataRequests.writeData(data_values);
    leak.getSeverity().setValue(0.0);// Turn off the leak
    pulse.processAction(leak);
    // Advance some time and print out the vitals
    pulse.advanceTime_s(5);
    // get the values of the data you requested at this time
    data_values = pulse.pullData();
    // And write it out to the console
    dataRequests.writeData(data_values);
  }

}
